internal/rotate.go

```diff
@@ -0,0 +1,88 @@
+package internal
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+)
+
+type CertRotationState struct {
+	EstServer string
+}
+
+type CertRotationHandler struct {
+	State           CertRotationState
+	stateFile       string
+	stateFileExists bool
+	app             *App
+	client          *http.Client
+	crypto          *EciesCrypto
+	steps           []CertRotationStep
+}
+
+type CertRotationStep interface {
+	Name() string
+	Completed(state CertRotationState) bool
+	Execute(handler *CertRotationHandler) error
+}
+
+func NewCertRotationHandler(app *App, stateFile, estServer string) CertRotationHandler {
+	client, crypto := createClient(app.sota)
+	handler := CertRotationHandler{
+		State:           CertRotationState{EstServer: estServer},
+		stateFile:       stateFile,
+		stateFileExists: true,
+		app:             app,
+		client:          client,
+		crypto:          crypto.(*EciesCrypto),
+		steps:           []CertRotationStep{},
+	}
+	bytes, err := os.ReadFile(stateFile)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			handler.stateFileExists = false
+		} else {
+			// Looks like we started a rotation, we should try and finish it
+			log.Printf("Error reading %s, return empty rotation state: %s", stateFile, err)
+		}
+		return handler
+	}
+	if err = json.Unmarshal(bytes, &handler.State); err != nil {
+		log.Printf("Error unmarshalling rotation state, return empty rotation state %s", err)
+		return handler
+	}
+	return handler
+}
+
+func (h *CertRotationHandler) StateFileExists() bool {
+	return h.stateFileExists
+}
+
+func (h *CertRotationHandler) Save() error {
+	bytes, err := json.Marshal(h.State)
+	if err != nil {
+		return err
+	}
+	return safeWrite(h.stateFile, bytes)
+}
+
+func (h *CertRotationHandler) Rotate() error {
+	for _, step := range h.steps {
+		if step.Completed(h.State) {
+			log.Printf("Step already completed: %s", step.Name())
+		} else {
+			log.Printf("Executing step: %s", step.Name())
+			err := step.Execute(h)
+			if saveErr := h.Save(); saveErr != nil {
+				return fmt.Errorf("Unable to save state: %w", saveErr)
+			}
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return os.Rename(h.stateFile, h.stateFile+".completed")
```

Comment:
Why do we need to rename it?
Maybe a current state can be represented by a field of the structure stored in the state file?

Reply:
It saves you from having to read/parse/analyze the file every time you run fioconfig. It's also kind of nice when visually looking at the directory to understand what's going on.

It also makes you not have to worry about backward compatibility of the file format. We could change the format or content anyway we want with this approach.

Reply:
The file still can be named `state.<state|step name>`. I think the rename in #87 can be removed and the rename should happen in #69, `os.Rename(h.stateFile, h.stateFile+"."+stepName`. And in, #79, h.Save(step.Name())

Reply:
on the final step, you could directly write `cert-rotation.state.completed` but that would mean `cert-rotation.state` didn't get updated. You'd need to call `unlink` on that and it wouldn't be transactionally safe.